Objects/frameobject.c

```diff
@@ -16,6 +16,622 @@
+        if (PyTuple_GET_ITEM(co->co_localsplusnames, i) == key ||
+                _PyUnicode_EQ(PyTuple_GET_ITEM(co->co_localsplusnames, i), key)) {
+            PyObject *value = framelocalsproxy_getval(frame, co, i);
+            if (value != NULL) {
+                return Py_NewRef(value);
+            }
+        }
+    }
+
+    // Okay not in the fast locals, try extra locals
+
+    PyObject *extra = frame->f_extra_locals;
+    if (extra != NULL) {
+        PyObject *value = PyDict_GetItem(extra, key);
+        if (value != NULL) {
+            return Py_NewRef(value);
+        }
+    }
+
+    PyErr_Format(PyExc_KeyError, "local variable '%R' is not defined", key);
+    return NULL;
+}
+
+static int
+framelocalsproxy_setitem(PyObject *self, PyObject *key, PyObject *value)
+{
+    /* Merge locals into fast locals */
+    PyFrameObject* frame = ((PyFrameLocalsProxyObject*)self)->frame;
+    PyObject** fast = _PyFrame_GetLocalsArray(frame->f_frame);
+    PyCodeObject* co = PyFrame_GetCode(frame);
+
+    if (value == NULL) {
+        PyErr_SetString(PyExc_TypeError, "cannot remove variables from FrameLocalsProxy");
+        return -1;
+    }
+
+    for (int i = 0; i < co->co_nlocalsplus; i++) {
+        if (PyTuple_GET_ITEM(co->co_localsplusnames, i) == key ||
+                _PyUnicode_EQ(PyTuple_GET_ITEM(co->co_localsplusnames, i), key)) {
+            _PyLocals_Kind kind = _PyLocals_GetKind(co->co_localspluskinds, i);
+
+            PyObject *oldvalue = fast[i];
+            PyObject *cell = NULL;
+            if (kind == CO_FAST_FREE) {
+                // The cell was set when the frame was created from
+                // the function's closure.
+                assert(oldvalue != NULL && PyCell_Check(oldvalue));
+                cell = oldvalue;
+            } else if (kind & CO_FAST_CELL && oldvalue != NULL) {
+                if (PyCell_Check(oldvalue)) {
+                    cell = oldvalue;
+                }
+            }
+            if (cell != NULL) {
+                oldvalue = PyCell_GET(cell);
+                if (value != oldvalue) {
+                    PyCell_SET(cell, Py_XNewRef(value));
+                    Py_XDECREF(oldvalue);
+                }
+            } else if (value != oldvalue) {
+                Py_XSETREF(fast[i], Py_NewRef(value));
+            }
+            Py_XDECREF(value);
+            return 0;
+        }
+    }
+
+    // Okay not in the fast locals, try extra locals
+
+    PyObject *extra = frame->f_extra_locals;
+
+    if (extra == NULL) {
+        extra = PyDict_New();
+        if (extra == NULL) {
+            return -1;
+        }
+        frame->f_extra_locals = extra;
+    }
+
+    assert(PyDict_Check(extra));
+
+    if (PyDict_SetItem(extra, key, value) < 0) {
+        return -1;
+    }
+
+    return 0;
+}
+
+static PyObject*
+framelocalsproxy_contains(PyObject *self, PyObject *key)
+{
+    PyObject* keys = framelocalsproxy_keys(self, NULL);
```

Comment:
If you factor out the key search, this function can become more efficient. Something like:
```C
int result = framelocalsproxy_keyindex(self, key);
if (result >= 0) {
    Py_RETURN_TRUE;
}
/* Need to distinguish between not found and an exception */
if (result == -1) {
    return NULL;
}
assert(result == NOT_FOUND);
if (self->f_extra_locals) {
    result= PyDict_Contains(self->f_extra_locals, key);
}
if (result< 0) {
    return NULL;
}
return return PyBool_FromLong(result);
}